packages/headless/source_docs/ssr-extend-engine-definitons.md

```diff
@@ -0,0 +1,235 @@
+3. It dispatches the `searchAction` given to it.
+4. It returns the engine and controllers it initialized.
+
+If you want to access the engine and controllers, you can use `build`, which effectively replaces steps 1 and 2.
+You can then use `hydrateStaticState.fromBuildResult`, which effectively replaces steps 3 and 4.
+
+```ts
+const { engine, controllers } = await engineDefinition.build();
+const staticState = await engineDefinition.hydrateStaticState.fromBuildResult({
+  buildResult: { engine, controllers },
+  searchAction,
+});
+```
+
+## Keep the server and client aligned
+
+If you choose to manipulate the engine or controllers before passing them to `fromBuildResult`, this could effect the state that’s returned by whichever `fromBuildResult` was called.
+For this reason, we recommend that you extract any manipulations you do to your engine into a separate function.
+You can then use it for both `fetchStaticState.fromBuildResult` and `hydrateStaticState.fromBuildResult`, as in the following code samples.
+
+In `common/engine-definition.ts`:
+
+```ts
+import {
+  defineSearchEngine,
+  defineSearchBox,
+  defineResultList,
+  defineFacet,
+  getSampleSearchEngineConfiguration,
+  loadQueryActions,
+  SearchCompletedAction,
+} from '@coveo/headless-react/ssr';
+
+const engineDefinition = defineSearchEngine({ ①
+  configuration: getSampleSearchEngineConfiguration(),
+  controllers: {
+    searchBox: defineSearchBox(),
+    resultList: defineResultList(),
+    authorFacet: defineFacet({ field: "author" }),
+    sourceFacet: defineFacet({ field: "source" }),
+  },
+});
+
+async function getBuildResult() { ②
+  const buildResult = await engineDefinition.build();
+  const { updateQuery } = loadQueryActions(buildResult.engine);
+  buildResult.engine.dispatch(updateQuery({ q: "I like trains" }));
+  return buildResult;
+}
+
+export async function fetchStaticState() { ③
+  return engineDefinition.fetchStaticState.fromBuildResult({
+    buildResult: await getBuildResult(),
+  });
+}
+
+export async function hydrateStaticState(options: { ④
+  searchAction: SearchCompletedAction;
+}) {
+  return engineDefinition.hydrateStaticState.fromBuildResult({
+    buildResult: await getBuildResult(),
+    searchAction: options.searchAction,
+  });
+}
+```
+
+1. Don’t export `engineDefinition`.
+2. Extract your logic to obtain the build result.
+3. Export your own customized version of `fetchStaticState`.
+4. Export your own customized version of `hydrateStaticState`.
+
+In `server.ts`:
+
+```ts
+import { fetchStaticState } from './common/engine-definition.ts';
+
+// ...
+
+const staticState = await fetchStaticState(); ①
+// ...
+```
+
+1. Your server code isn’t polluted with with a lot of complex logic.
+
+In `client.ts`:
+
+```ts
+import { hydrateStaticState } from './common/engine-definition.ts';
+
+// ...
+
+const hydratedState = await hydrateStaticState({
+  searchAction: staticState.searchAction,
+});
+```
+
+<dl><dt><strong>⚠️ WARNING</strong></dt><dd>
```

Comment:
It would be nice to have some kind of box for this. Right now, with indentation alone, it's not immediately clear what's part of the warning:
<img width="992" height="831" alt="Screenshot 2025-11-26 at 12 58 00 PM" src="https://github.com/user-attachments/assets/b376fd7d-a910-4fd3-bdb8-c6e43012f821" />